File: Java/longest-palindromic-substring.java
package pers.kindem.leetcode.java.solution.LongestPalindromicSubstring5;

public class Solution {
    public String longestPalindrome(String s) {
        int length = 0;
        String ans = "";

        int p, q;
        // 遍历所有字符，分别以他们作为核，看最大能生长到多长
        for (int i = 0; i < s.length(); i++) {
            p = i - 1;
            q = i + 1;
            while (p >= 0 && q < s.length()) {
                if (s.charAt(p) == s.charAt(q)) {
                    p--;
                    q++;
                } else break;
            }
            if (length < q - p - 1) {
                length = q - p - 1;
                ans = s.substring(p + 1, q);
            }
        }
        // 两个字符为单位作为核，再一次看最大能生长到多长
        for (int i = 0; i < s.length() - 1; i++) {
            p = i;
            q = i + 1;
            while (p >= 0 && q < s.length()) {
                if (s.charAt(p) == s.charAt(q)) {
                    p--;
                    q++;
                } else break;
            }
            if (length < q - p - 1) {
                length = q - p - 1;
                ans = s.substring(p + 1, q);
            }
        }
        return ans;
    }
}

File: Python/longest-palindromic-substring.py
class Solution:
    def __checkIfNextPalindrome(self, s, p, q, count):
        if p - 1 < 0 or q + 1 > len(s) - 1:
            return [p, q, count]
        else:
            if s[p - 1] == s[q + 1]:
                return self.__checkIfNextPalindrome(s, p - 1, q + 1, count + 2)
            else:
                return [p, q, count]

    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        ans = ''
        max = 0
        p = 0
        while p < len(s):
            q = p
            while q < len(s) and s[p] == s[q]:
                q += 1
            q -= 1
            tmp = self.__checkIfNextPalindrome(s, p, q, q + 1 - p)
            if tmp[2] > max:
                max = tmp[2]
                ans = s[tmp[0]:tmp[1] + 1]
            p += 1
        return ans

